models/monai_prostate158/utils/Prostate158Runner.py

```diff
@@ -0,0 +1,85 @@
+MHub - MONAI Prostate158 Runner
+-------------------------------------------------
+
+-------------------------------------------------
+Author: Cosmin Ciausu
+Email:  cciausu97@gmail.com
+-------------------------------------------------
+"""
+# TODO: support multi-i/o and batch processing on multiple instances
+
+from typing import List, Optional
+import os, subprocess, shutil, glob, sys
+import SimpleITK as sitk, numpy as np
+from mhubio.core import Module, Instance, InstanceData, DataType, FileType, IO
+from mhubio.modules.runner.ModelRunner import ModelRunner
+import json
+
+@IO.Config('apply_center_crop', bool, False, the='flag to apply center cropping to input_image')
+
+class Prostate158Runner(ModelRunner):
+
+    apply_center_crop : bool
+
+    @IO.Instance()
+    @IO.Input("in_data", 'nifti:mod=mr', the="input T2 sequence data to run prostate158 on")
+    @IO.Output('out_data', 'monai_prostate158.nii.gz', 
+               'nifti:mod=seg:model=Prostate158:roi=PROSTATE_TRANSITION_ZONE,PROSTATE_PERIPHERAL_ZONE', 
+               data='in_data', bundle='model', the="predicted segmentation model")
+
+    def task(self, instance: Instance, in_data: InstanceData, out_data: InstanceData) -> None:
+
+        # bring input data in nnunet specific format
+        # NOTE: only for nifti data as we hardcode the nnunet-formatted-filename (and extension) for now.
+        assert in_data.type.ftype == FileType.NIFTI
+        assert in_data.abspath.endswith('.nii.gz')
+        datalist = [in_data.abspath]
+
+        if self.apply_center_crop:
+            in_dir_cropped = self.config.data.requestTempDir(label="monai-crop-in")
+            in_data_processed = os.path.join(in_dir_cropped, "image_cropped.nii.gz")
+            self.subprocess([sys.executable, f"{os.path.join(os.environ['BUNDLE_ROOT'], 'scripts', 'center_crop.py')}",
+            "--file_name", in_data.abspath, "--out_name",in_data_processed], text=True)
+            datalist = [in_data_processed]
+
+        # define output folder (temp dir) and also override environment variable for nnunet
+        out_dir = self.config.data.requestTempDir(label="monai-model-out")
+       
+        bash_command = [sys.executable, 
+        "-m", "monai.bundle", "run", "evaluating"]
+        bash_command += ["--meta_file", os.path.join(os.environ['BUNDLE_ROOT'], "configs", "metadata.json")]
+        bash_command += ["--config_file", os.path.join(os.environ['BUNDLE_ROOT'], "configs", "inference.json")]
+        bash_command += ["--datalist", str(datalist)]
+        bash_command += ["--output_dir", out_dir]
+        bash_command += ["--bundle_root", os.environ['BUNDLE_ROOT']]
+        bash_command += ["--dataloader#num_workers", "0"]
+        print(bash_command)
+        self.subprocess(bash_command, text=True)
+
+        # get output data
+        out_path = glob.glob(os.path.join(out_dir, "**", 
+        "*.nii.gz"), recursive=True)[0]
+
+        if self.apply_center_crop:
```

Comment:
Why do we have a script (`center_crop.py`)for preprocessing and a custom implementation for postprocessing?

Reply:
center_crop.py comes from the MONAI bundle when building the docker image, whereas the postprocessing is done by me only, necessary to pad the resulting seg output to the original image space, right before conversion to SEG DICOM.  To make it 'cleaner', the only way I see is to take out the center_crop.py out of the monai bundle and add it to mhub as a utility script, alongside the postprocessing (line 65+). Not sure though if it is much better. 

FYI regarding the post-processing : Another strategy would have been to use the cropped image as a reference and then no need for this post-processing code, however I personally did not like it since it modifies the original image and padding with 0 the resulting seg image is not affecting it.

Reply:
Thx for the explaination. That's just fine then.